### PR TITLE
Cvae

### PR DESCRIPTION
**Pull request by ZhihongShao**
**Description:**

Some details (e.g., number of samples -- args.repeat_N, which smooth function to use for BLEU) need to be specified for multi-reference test.

**Reference Issues:** #11

### CODE REVIEW

File: contk/dataloader/multi_turn_dialog.py [hunk 1]
```diff
@@ -9,6 +10,7 @@
 from .dataloader import BasicLanguageGeneration
 from ..metric import MetricChain, MultiTurnPerplexityMetric, MultiTurnBleuCorpusMetric, \
 	MultiTurnDialogRecorder
+from ..metric import bleu_prec_rec_score_fn, cos_sim_prec_rec_score_fn, PrecisionRecallMetric
```
Comment (hzhwcmhf) [inline contk/dataloader/multi_turn_dialog.py]: Use name like ``BleuPrecRecScore``

File: tests/dataloader/dummy_switchboardcorpus/switchboard_corpus_dev.jsonl [hunk 2]
```diff
@@ -0,0 +1,10 @@
...e", ["acknowledge_(backchannel)", [0.0, 0.0, 1.0, 0.3182]]], ["B", "and he he's real fun to listen to", ["statement-non-opinion", [0.0, 0.68, 0.32, 0.5106]]], ["A", "uh-huh", ["acknowledge_(backchannel)", [0.0, 1.0, 0.0, 0.0]]], ["B", "bonnie rait i like her a lot in fact when ever she won her big grammy i think there was like five of them or something was it five or", ["statement-non-opinion", [0.0, 0.734, 0.266, 0.8271]]], ["A", "um-hum", ["abandoned_or_turn-exit/uninterpretable", [0.0, 1.0, 0.0, 0.0]]], ["A", "um-hum", ["abandoned_or_turn-exit/uninterpretable", [0.0, 1.0, 0.0, 0.0]]], ["A", "uh-huh", ["acknowledge_(backchannel)", [0.0, 1.0, 0.0, 0.0]]], ["A", "yeah something like that it was last year wasn't it", ["statement-non-opinion", [0.0, 0.63, 0.37, 0.5719]]], ["B", "yeah last year or the year before i can't remember but i was listening to a particular radio station and the one dj said who the heck is bonnie rait and why does she deserve five grammies", ["statement-non-opinion", [0.0, 0.954, 0.046, 0.1531]]], ["B", "and i vowed then that i was not gonna listen to that radio station any more if that dj could make such an", ["statement-non-opinion", [0.0, 1.0, 0.0, 0.0]]]]}
```
Comment (hzhwcmhf) [inline tests/dataloader/dummy_switchboardcorpus/switchboard_corpus_dev.jsonl]: use a smaller dummy file

**Comment by hzhwcmhf** [comment]
metrics should not introduce function api.

https://github.com/thu-coai/contk/blob/2db574a2dc0c061c85bb0278186f235fc8284ad8/contk/metric/metric.py#L17-L150

Now you have `bleu_prec_rec_score_fn`, `cos_sim_prec_rec_score_fn`, `PrecisionRecallMetric`

Please change it to `BleuPrecisionRecallMetric`, `EmbSimilarityPrecisionRecallMetric`.
You can implement your function as a local function, and expose Metric class to user.
(local function start with underline  like `_PrecisionRecallMetric`)

File: contk/metric/metric.py [hunk 3]
```diff
@@ -84,7 +220,6 @@ def forward(self, data):
 
 	def close(self):
 		'''Return a dict which contains:
-
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: Don't change line breaks

File: contk/dataloader/multi_turn_dialog.py [hunk 4]
```diff
@@ -288,3 +291,391 @@ def _load_data(self):
...
+	            max_sen_length (int): All sentences longer than `max_sen_length` will be shortened
+	                to first `max_sen_length` tokens. Default: 50.
+	            session_window (int): The maximum number utterance kept in a session. Default: 10
+
+	        Refer to :class:`.MultiTurnDialog` for attributes and methods.
+
+	        Todo:
+	            * add references
+	        '''
+	def __init__(self, file_path, max_vocab_size=10000, max_sen_length=50, session_window=10):
```
Comment (hzhwcmhf) [inline contk/dataloader/multi_turn_dialog.py]: please use similiar api like `UbuntuCorpus`. Use `min_vocab_times` instead of `max_vocab_size`, `max_turn_length` instead of `session_window`

Comment (hzhwcmhf) [reply contk/dataloader/multi_turn_dialog.py]: Moreover, there should be an ``invalid_vocab_times``. This is a new feature of dataloader. Simply, **valid words** are for dataloader, **invalid words** are for metrics (but haven't been merge to master now.). You can see #63 for details. So you have to change your `_load_data` to be compatible with the new feature.

File: contk/dataloader/multi_turn_dialog.py [hunk 5]
```diff
@@ -288,3 +291,391 @@ def _load_data(self):
...
+		'''Get a batch of specified `index`.
+
+		Arguments:
+			key (str): must be contained in `key_name`
+			index (list): a list of specified index
+
+		Returns:
+			(dict): A dict at least contains:
+
+				* session (dict): Return of MultiTurnDialog.get_batch
```
Comment (hzhwcmhf) [inline contk/dataloader/multi_turn_dialog.py]: please don't make another dict there, just let the contents in `session` be items of `res`

File: contk/dataloader/multi_turn_dialog.py [hunk 6]
```diff
@@ -288,3 +291,391 @@ def _load_data(self):
...
+				* topics (list): A 1-d list, the topic ids of sessions
+					Size: ^[batch_size]^
+				* A_profiles (list): A 2-d list, the profiles (a 1-d list with size 4) of speaker ^A^.
+					Size: ^[batch_size, 4]^
+				* B_profiles (list): A 2-d list, the profiles (a 1-d list with size 4) of speaker ^B^.
+					Size: ^[batch_size, 4]^
+				* floors (list): A 2-d list, speakers of utterances. The inner lists may have different length.
```
Comment (hzhwcmhf) [inline contk/dataloader/multi_turn_dialog.py]: I have known the format of switchboard is different from ubuntu, where `floors` indicates which people is speaking, while `ubuntu` will change the speaker each turn. But we must remain the same behavior in all dataloaders, so we should preprocessing the data. For example:

``<go> hello <eos>`` and ``<go> how are you <eos>`` from speaker A should be merged into one turn like:
``<go> hello <eos> how are you <eos> <eot>``

File: contk/dataloader/multi_turn_dialog.py [hunk 7]
```diff
@@ -288,3 +291,391 @@ def _load_data(self):
...
+			if end - 1 < utt_len:
+				raw_batch['responses_length'].append( \
+					batch_data['session']['sent_length'][i][end - 1])
+			else:
+				raw_batch['responses_length'].append(1)
+		raw_batch['contexts'] = batch_data['session']['sent'][:, start: end - 1]
+		raw_batch['responses'] = batch_data['session']['sent'][:, end - 1]
+		return raw_batch
+
+	def split_session(self, batch_data, inference=False):
```
Comment (hzhwcmhf) [inline contk/dataloader/multi_turn_dialog.py]: Move it to your model, if you just want to use it. I will consider to make a api like this, but not now.

File: contk/dataloader/multi_turn_dialog.py [hunk 8]
```diff
@@ -288,3 +291,391 @@ def _load_data(self):
...
+				test_data['session'][idx].append(\
+					[self.go_id, self.word2id['<d>'], self.eos_id])
+				test_data['dialog_acts'][idx].append(0)
+				utt2id = lambda utt: list(map(lambda w: \
+								self.word2id.get(w, self.unk_id), utt[1].strip().split()))
+				test_data['candidates'].append(list(map(utt2id, line['responses'])))
+				idx += 1
+		return test_data
+
+	def multi_reference_batches(self, batch_size):
```
Comment (hzhwcmhf) [inline contk/dataloader/multi_turn_dialog.py]: Same like ``split_session``

**Comment by hzhwcmhf** [comment]
You may write a unit test for your dataloader like this

https://github.com/thu-coai/contk/blob/3e072fcf6f88342bfb23527e2800ac9c597129ba/tests/dataloader/test_multi_turn_dialog.py#L212-L243

File: tests/dataloader/dummy_switchboardcorpus/switchboard_corpus_dev.jsonl [hunk 9]
```diff
@@ -0,0 +1,10 @@
... 1.0, 0.296]]], ["B", "that's true", ["agree/accept", [0.0, 0.263, 0.737, 0.4215]]], ["B", "yeah oh my atlanta braves", ["agree/accept", [0.0, 0.37, 0.63, 0.6249]]], ["A", "well they did well", ["abandoned_or_turn-exit/uninterpretable", [0.0, 0.323, 0.677, 0.4939]]], ["A", "it's yeah", ["acknowledge_(backchannel)", [0.0, 0.313, 0.688, 0.296]]], ["A", "it is it is and that's why i've i guess my mind went blank it's like last season was so long ago and i haven't really even started thinking about basketball yet so", ["statement-non-opinion", [0.0, 0.923, 0.077, 0.3612]]], ["B", "yeah", ["acknowledge_(backchannel)", [0.0, 0.0, 1.0, 0.296]]], ["A", "but this wasn't a good subject but we'll see how our predictions come true", ["statement-opinion", [0.175, 0.618, 0.208, 0.1508]]], ["B", "it's a little early", ["statement-opinion", [0.0, 1.0, 0.0, 0.0]]], ["B", "we definitely will", ["agree/accept", [0.0, 0.426, 0.574, 0.4019]]], ["A", "okay it's good talking to you", ["conventional-closing", [0.0, 0.455, 0.545, 0.5859]]], ["B", "good to talk to you bye-bye", ["conventional-closing", [0.0, 0.633, 0.367, 0.4404]]], ["A", "alrighty bye-bye", ["conventional-closing", [0.0, 1.0, 0.0, 0.0]]]]}
```
Comment (hzhwcmhf) [inline tests/dataloader/dummy_switchboardcorpus/switchboard_corpus_dev.jsonl]: use a smaller dummy file

File: contk/dataloader/multi_turn_dialog.py [hunk 10]
```diff
@@ -294,3 +297,193 @@ def _load_data(self):
...
+
+	def get_batch(self, key, index):
+		'''Get a batch of specified `index`.
+
+		Arguments:
+			key (str): must be contained in `key_name`
+			index (list): a list of specified index
+
+		Returns:
+			(dict): A dict at least contains what is in the return of MultiTurnDialog.get_batch
```
Comment (hzhwcmhf) [inline contk/dataloader/multi_turn_dialog.py]: Don't use `at least`. List all the items besides what returned by MultiTurnDialog.get_batch

File: contk/dataloader/multi_turn_dialog.py [hunk 11]
```diff
@@ -294,3 +297,193 @@ def _load_data(self):
...
+		'''Get metric for precision and recall in terms of BLEU, cosine similarity.
+
+        It contains:
+
+        * :class:`.metric.PrecisionRecallMetric`
+
+        Arguments:
+            embed (:class:`numpy.array`): Glove word embedding
+        '''
+		metrics = {}
```
Comment (hzhwcmhf) [inline contk/dataloader/multi_turn_dialog.py]: Use MetricChain

**Comment by hzhwcmhf** [comment]
* Fix the errors in travis-CI
* Add a test like ``contk/models/seq2seq-tensorflow/test_seq2seq_tensorflow.py``

**Comment by ZhihongShao** [comment]
MultiTurnDialogRecorder removes generated texts that have no valid words, which leads to the error in travis CI.

**Comment by hzhwcmhf** [comment]
I'll check it soon @ZhihongShao

**Comment by hzhwcmhf** [comment]
I've fix the problem. Pull master and check it @ZhihongShao

File: contk/metric/metric.py [hunk 12]
```diff
@@ -6,13 +6,155 @@
...
+		self.res_prefix = 'BLEU-{}'.format(ngram)
+
+	def score(self, gen, reference):
+		r'''Score_fn of BLEU-ngram precision and recall
+
+        Returns:
+            (scalar): sentence bleu score \in [0, 1]
+        '''
+		weights = [0] * 4
+		weights[self.ngram - 1] = 1
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]:
```
    The default BLEU calculates a score for up to 4-grams using uniform
    weights (this is called BLEU-4). To evaluate your translations with
    higher/lower order ngrams, use customized weights. E.g. when accounting
    for up to 5-grams with uniform weights (this is called BLEU-5) use:

    >>> weights = (1./5., 1./5., 1./5., 1./5., 1./5.)
```
refer to https://www.nltk.org/_modules/nltk/translate/bleu_score.html

File: models/CVAE-tensorflow/Readme.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+	resp:	hum okay well i mean also along the same lines i know that there's somebody i read one time about somebody's developed a test to look at the emissions from your car <eos>
+	gen:	um-hum <eos> yeah <eos> well what do you think of it is that you know <eos>
+
+	post:	hum okay well i mean also along the same lines i know that there's somebody i read one time about somebody's developed a test to look at the emissions from your car <eos>
+	resp:	uh-huh <eos> right <eos>
+	gen:	uh-huh <eos>
+
+...
+
+Test Multi Reference Result:
```
Comment (hzhwcmhf) [inline models/CVAE-tensorflow/Readme.md]: move the results below to Performance

**Comment by ZhihongShao** [comment]
Index out of range error in `test_process` of hred model.

File: models/CVAE-tensorflow/Readme.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+#### For developer
+
+- You should remain similar output in this task.
+
+### Performance
+
+|               | SwitchboardCorpus |
+| ------------- | ----------------- |
+| Perplexity (KL) | 21.70 (12.20) |
```
Comment (hzhwcmhf) [inline models/CVAE-tensorflow/Readme.md]: I meant that put the performance here. Including BLEU-n precision/recall and other metrics.

**Comment by hzhwcmhf** [comment]
@ZhihongShao  Pull master for updated hred test
